src/java_bytecode/java_bytecode_parser.cpp

```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
+    {
+      u2 skipped_argument = u2_values[i];
+      recognized &= pool_entry(skipped_argument).tag == CONSTANT_Integer;
     }
-    else
+
+    if(!recognized)
     {
-      lambda_method_handlet lambda_method_handle;
-      lambda_method_handle.handle_type = method_handle_typet::UNKNOWN_HANDLE;
-      lambda_method_handle.u2_values = std::move(u2_values);
-      parsed_class.lambda_method_handle_map[{parsed_class.name, i}] =
-        lambda_method_handle;
-      error() << "ERROR: num_bootstrap_arguments must be at least 3" << eom;
+      debug() << "format of BootstrapMethods entry not recognized: extra "
+                 "arguments of wrong type"
+              << eom;
+      store_unknown_method_handle(
+        parsed_class, bootstrap_method_index, std::move(u2_values));
+      continue;
+    }
+
+    const pool_entryt &interface_type_argument =
+      pool_entry(interface_type_index);
+    const pool_entryt &method_handle_argument = pool_entry(method_handle_index);
+    const pool_entryt &method_type_argument = pool_entry(method_type_index);
+
+    if(
+      !(interface_type_argument.tag == CONSTANT_MethodType &&
+        method_handle_argument.tag == CONSTANT_MethodHandle &&
+        method_type_argument.tag == CONSTANT_MethodType))
+    {
+      debug() << "format of BootstrapMethods entry not recognized: arguments "
+                 "wrong type"
+              << eom;
+      store_unknown_method_handle(
+        parsed_class, bootstrap_method_index, std::move(u2_values));
+      continue;
+    }
+
+    debug() << "INFO: parse lambda handle" << eom;
+    optionalt<lambda_method_handlet> lambda_method_handle =
+      parse_method_handle(method_handle_infot{method_handle_argument});
+
+    if(!lambda_method_handle.has_value())
+    {
+      debug() << "format of BootstrapMethods entry not recognized: method "
+                 "handle not recognised"
+              << eom;
+      store_unknown_method_handle(
+        parsed_class, bootstrap_method_index, std::move(u2_values));
+      continue;
     }
+
+    // If parse_method_handle can't parse the lambda method, it should return {}
+    POSTCONDITION(
+      lambda_method_handle->handle_type !=
+      method_handle_typet::UNKNOWN_HANDLE);
+
+    lambda_method_handle->interface_type =
+      pool_entry(interface_type_argument.ref1).s;
+    lambda_method_handle->method_type =
+      pool_entry(method_type_argument.ref1).s;
+    lambda_method_handle->u2_values = std::move(u2_values);
+    debug() << "lambda function reference "
+            << id2string(lambda_method_handle->lambda_method_name)
+            << " in class \"" << parsed_class.name << "\""
+            << "\n  interface type is "
+            << id2string(pool_entry(interface_type_argument.ref1).s)
+            << "\n  method type is "
+            << id2string(pool_entry(method_type_argument.ref1).s)
+            << eom;
+    parsed_class.lambda_method_handle_map[{parsed_class.name, bootstrap_method_index}] =
```

Comment:
Nit: Don't we have an 80-char limit?

Reply:
The way we access this map looks a bit unnatural. Why does name of `parsed_class` have to be an element of the map entry, that itself belongs to the same class. Obviously the name will always be identical for a given `parsed_class` or am I missing something? Could we have a wrapper method in `java_bytecode_parse_treet::classt` that cares about adding an element to the map instead of accessing the map directly?

Reply:
Yes I agree  tagging @mgudemann to specifically review what I've done here but I agree this info is redundant and a method will make updating this in the future simpler - will do.